amaranth/back/rtlil.py

```diff
@@ -831,7 +831,11 @@ def _convert_fragment(builder, fragment, name_map, hierarchy):
 
     if isinstance(fragment, mem.MemoryInstance):
         memory = fragment.memory
-        init = "".join(format(ast.Const(elem, ast.unsigned(memory.width)).value, f"0{memory.width}b") for elem in reversed(memory.init))
+        if isinstance(memory.shape, ast.ShapeCastable):
```

Comment:
please move the const-casting code to `Memory.__init__`, so that 1) it's deduplicated between RTLIL and sim, 2) any errors are reported earlier, during `Memory` construction

Reply:
er, to `Memory.init` setter I guess

Reply:
I considered doing it that way, but I figure either way we should keep the uncasted values around so the getter is consistent with the setter. The getter also returns a mutable list, so the following code is valid:
```py
mem = Memory(width = 8, depth = 16)
for i in range(mem.depth):
    mem.init.append(…)
```
If we want to do the cast in the setter, we should also make the list immutable without going through the setter.

Reply:
We can make `MemoryInitializer` a new `MutableSequence` collection that ensures all of the values inside are const-castable.